Hardwax parser

TEdwardJ opened the pull request:
- VinylParser implementation added for hardwax.com;
- related changes in entities, such as Offer and RawOffer;

TEdwardJ committed:
Hardwax parser:
 - HardwaxParser added;
 - tests for the parser added;
 - Refactoring for DetailedVinylParser and its implementation - getPriceDetailsFromDocument() method added; it is used in getPriceFromDocument() and getOptionalCurrencyFromDocument() methods;
 - new parser is included into start configuration in Strarter class;
 - RawOfferSorter has the issue with matching rate evaluation fixed;

TEdwardJ committed:
Hardwax parser:
 - HardwaxParser - changes in the parser configuration;
 - Offer and RawOffer entities has genre field excluded from equals and hashCode methods;

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:
```diff
@@ -293,6 +298,11 @@ public double getPriceFromDocument(Element document) {
+        @Override
+        public List<String> getPriceDetailsFromDocument(Element document) {
+            return document.select(PRICE_DETAILS_SELECTOR).eachText();
+        }
```
Is it the same method as 221-224 rows?

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/util/impl/HardWaxParser.java:
```diff
@@ -0,0 +1,229 @@
+@Slf4j
+public class HardWaxParser extends VinylParser {
+
+    protected static final String BASE_LINK = "https://hardwax.com";
```
why protected?

TEdwardJ replied:
since it used in tests

Yuliia-Sh replied:
for test package private is enough

AlexKievDeveloper commented on src/test/java/com/vinylteam/vinyl/util/impl/HardWaxParserTest.java:
```diff
@@ -0,0 +1,89 @@
+class HardWaxParserTest {
+
+    private HardWaxParser parser;
+
```
Please remove empty line

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/entity/Offer.java:
```diff
@@ -107,12 +107,12 @@ public boolean equals(Object o) {
-        return Objects.hash(id, uniqueVinylId, shopId, price, currency, genre, catNumber, inStock, offerLink);
+        return Objects.hash(id, uniqueVinylId, shopId, price, currency, catNumber, inStock, offerLink);
     }
 
```
why to not use Lombok? Won't code more readable?

TEdwardJ replied:
Let`s try, we are always able to rollback changes in case it will not appear as readable as expected

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/util/PriceUtils.java:
```diff
@@ -16,10 +16,10 @@ public static double getPriceFromString(String fullPriceDetails) {
             } catch (Exception e) {
-                log.error("Error while getting price from price details from link {'priceDetails':{}, 'link':{}}", fullPriceDetails);
+                log.error("Error while getting price from price details from link {'priceDetails':{}}", fullPriceDetails);
             }
```
will be exception in the log?

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/util/impl/DetailedVinylParserImpl.java:
```diff
@@ -16,7 +16,7 @@
 
-    private final ParserConfiguration config;
+    final ParserConfiguration config;
 
```
why changed access modifier?

TEdwardJ replied:
to make it visible from descendants, but seems protected is more applicable

TEdwardJ committed:
Hardwax parser:
 - code style remarks;
 - CloneNlParser changed in order to use parser configuration rather than some additional implementations of DetailedVinylParser;
 - wrong test data for RawOffersSorter changed;

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/util/impl/HardWaxParser.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    class HardWaxDetailedParserImpl extends DetailedVinylParserImpl {
+
```
Why inner class? Shouldn't it be static?

Yuliia-Sh replied:
private?

TEdwardJ replied:
It is inner since this implementation is specific for hardwax.com parser and used only inside the parser.
it was supposed that an instance of the hardwax parser only will create instances of this inner class. The rest of parser tasks are covered by general implementation  - DetailedVinylParserImpl.
So the private modifier is a good idea, thanks!

TEdwardJ committed:
Hardwax parser review remarks:
 - Offer class is simplified with Lombok;
 - changes in access modifiers;
 - HardWaxParser has its inner class
 - PriceUtils shows exception in log;

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/util/impl/DetailedVinylParserImpl.java:
```diff
@@ -16,7 +16,7 @@
 public class DetailedVinylParserImpl implements DetailedVinylParser {
 
-    private final ParserConfiguration config;
+    protected final ParserConfiguration config;
```
Why protected? You will need access to this object from inheritor which is located in different package? If not, please use private or package-private(default) access modifier
